src/ast/ast_validations.c

```diff
@@ -1169,6 +1180,101 @@ static AST_Validation _ValidateMaps(const cypher_astnode_t *root, char **reason)
 	return AST_VALID;
 }
 
+/* validate list usage in subscript is correct */
+static AST_Validation _validateList(const cypher_astnode_t *root, char **reason) {
+	const cypher_astnode_type_t type = cypher_astnode_type(root);
+	// check that the operation returns a list - ragne function
+	if(type == CYPHER_AST_APPLY_OPERATOR) {
+		const cypher_astnode_t *funcNode =  cypher_ast_apply_operator_get_func_name(root);
+		const char *funcName = cypher_ast_function_name_get_value(funcNode);
+		// function name is NOT range
+		if(strcasecmp(funcName, "range")) {
+			asprintf(reason, "subscript index access expects range funcition; encountered '%s'", funcName);
+			return AST_INVALID;
+		}
+		// validate the number of arguments in range function 2-3
+		uint narguments = cypher_ast_apply_operator_narguments(root);
+		if(narguments < 2 || narguments > 3) {
+			asprintf(reason, "range function expects 2 or 3 argumetns; encountered %d", narguments);
+			return AST_INVALID;
+		}
+		// validate that all the arguments are integers
+		for(uint i = 0; i < narguments; i ++) {
+			const cypher_astnode_t *argument = cypher_ast_apply_operator_get_argument(root, i);
+			const cypher_astnode_type_t argument_type = cypher_astnode_type(argument);
+			if(argument_type != CYPHER_AST_INTEGER || argument_type != CYPHER_AST_IDENTIFIER) {
+				asprintf(reason, "expected interger or identifier; encountered %s",
+						 cypher_astnode_typestr(argument_type));
+				return AST_INVALID;
+			}
+		}
+	} else if(type != CYPHER_AST_COLLECTION && type != CYPHER_AST_IDENTIFIER) {
+		// list is a collection or identifier
+		// TODO: in current state, the identifier type is evluated in query runtime
+		// check if possible to evluate during ast validation
```

Comment:
Unfortunately, there is no connection between multiple uses of the same identifier in the parser-returned AST.

Reply:
Lets talk about it on Monday